### PR TITLE
side_discontinuous option for Exodus output

### PR DESCRIPTION
**Pull request by roystgnr**
## Reason
MOOSE has had support for `SIDE_DISCONTINUOUS` variables for a while, but no way to visualize them easily.  This should fix that.

## Design
By setting side_discontinuous=true in Exodus output, disconnected fictitious side elements are added to that output, on which `SIDE_DISCONTINUOUS` values are plotted, with the discontinuities preserved.

A couple new DGKernel subclasses, HFEMTestJump and HFEMTrialJump, let us test this using a simple HFEM problem, solved with `SIDE_DISCONTINUOUS` lagrange multiplier rather than with the non-fictitious side element option previously available.

## Impact
Additional capabilities for Exodus.

DGKernels now "count" when our error checking tests to make sure that every variable in the system has an associated kernel.

Refs #20191.  It won't close that, though; we still have more 3D element types to add support for.

There's a bit of copy-and-paste code I still need to refactor into a new base class for SideObject+InternalSideObject, but I'm out of time for today and I'd like to find out whether CI agrees with my workstation that this is passing tests.

### CODE REVIEW

**Comment by lindsayad** [comment]
I think those nonsparse failures are likely real failures

**Comment by roystgnr** [comment]
> I think those nonsparse failures are likely real failures

After half a day of trying to figure this out: they're worse than real failures.  Real failures mean I wrote code that didn't do what I meant, and need to write code that does what I meant instead.  These failures are failures of *definition*, where I didn't actually have a straightforward idea of what I meant.

DGKernel calculations appear to use "which element has the higher id" to decide which side of an element counts as *the* element vs the mere *neighbor* ... which would mean that as soon as we add a few processors to a distributed run and DistributedMesh helpfully reshuffles ids to give us contiguous id blocks on each rank, the definition of "element" vs "neighbor" inverts, and the definition of a Lagrange multiplier variable constraining directional jumps from element to neighbor also inverts, and exodiff starts whining at me about having rightfully noticed that `-1/6` and `1/6` are two different numbers.

I guess the right thing to do here is something like a `vertex_average()`-sorting-dependent sign flip in the new HFEM kernels, to make the results depend on geometry without regard for numbering.

I bet the old HFEM kernels with the non-fictitious side elements have the same problem, but never trigger it because they don't run distributed...

**Comment by lindsayad** [comment]
I would contend that difference only by sign in a Lagrange multiplier is not a true failure at all. What I would do is create a custom comparison that doesn't exodiff based on the LMs ... but then again I guess the whole point of this PR is the exodus output of these LMs :laughing: Would it be possible to pass the LM field through an aux kernel that takes the absolute value of the LM and then only write out that field?

**Comment by roystgnr** [comment]
> I guess the whole point of this PR is the exodus output of these LMs

Admittedly I do feel like I've gone down the rabbit hole here.  I haven't bothered to push anything until it's all fixed, but as part of fixing those doc failures I've written a test case to add to the documentation of a base class I refactored to properly support the postprocessor I wrote for a *different* test case ...

But I do want this stuff to be useful too, and if I can get this working (my quick fix idea seems to have worked perfectly in 1D but not at all in 2D???) then making it easier for future kernels to be tested via exodiff will be a valuable thing to have too.

Plus I really want to understand what I'm still doing wrong.

**Comment by roystgnr** [comment]
Okay, I get it now.  After making the orientation of lambda depend on geometry independent of numbering, I now get consistent lambda values on all side elements, and I was already getting consistent u values on all interiors ... but I'm also plotting u on those side elements.  And in this test u is discontinuous.  And I plot its value on a side exactly once, from whichever interior I happen to hit first when I'm plotting that side.  So if renumbering means I hit interiors in different orders, then I can plot the same internal side from different elements in different partitionings, and it can take on different values.

Technically u is discontinuous, IMHO it would be fair to just plot it as 0 or NaN on sides ... but users might find it more useful to plot it as the average of the values between the two interiors, and that's pretty much what we already do when plotting discontinuous fields at shared nodes in other contexts, too.

File: framework/src/systems/NonlinearSystemBase.C [hunk 1]
```diff
@@ -3340,6 +3340,7 @@ NonlinearSystemBase::checkKernelCoverage(const std::set<SubdomainID> & mesh_subd
   global_kernels_exist |= _nodal_kernels.hasActiveObjects();
 
   _kernels.subdomainsCovered(input_subdomains, kernel_variables);
+  _dg_kernels.subdomainsCovered(input_subdomains, kernel_variables);
```
Comment (lindsayad) [inline framework/src/systems/NonlinearSystemBase.C:3343]: nice

File: test/tests/outputs/exodus/exodus_side_discontinuous.i [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    type = ElementSidesL2Norm
+    variable = lambda
+  []
+[]
+
+[Executioner]
+  type = Steady
+  solve_type = 'NEWTON'
+  petsc_options_iname = '-pc_type -snes_linesearch_type -pc_factor_mat_solver_type'
+  petsc_options_value = 'lu       basic                 mumps'
```
Comment (lindsayad) [inline test/tests/outputs/exodus/exodus_side_discontinuous.i:78]: Not saying you need to but would be cool to use a field split here

**Review by lindsayad — CHANGES_REQUESTED** [review]
I'd like to see a `PetscJacobianTester` test to verify that the Jacobian for this higher-d-variable/side-variable coupling is accurate

**Comment by lindsayad** [comment]
Is there a way to hit `InternalSideUserObject::getFaceInfos`?

File: framework/doc/content/source/postprocessors/ElementSidesL2Norm.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+Where here $\Gamma$ is defined as the union of all sides of all
+elements in a mesh, counting internal sides shared by two elements
+once rather than twice.
+
+The side $L_2$ norm of a variable $u$ is computed as:
+
+!equation
+||u||_{L_2(\Gamma)} = \sqrt{\sum_{\text{mesh element sides}} \sum_{\text{quadrature point qp}} J[qp] W[qp] c[qp] u(qp)^2}
+
+These terms are initialized on eeach side, so the $J$ term accounts
```
Comment (lindsayad) [inline framework/doc/content/source/postprocessors/ElementSidesL2Norm.md:19]:
```suggestion
These terms are initialized on each side, so the $J$ term accounts
```

**Comment by lindsayad** [comment]
Yea probably a new distinct test would be the easiest. Let me know if you'd like any help cooking one up

**Comment by roystgnr** [comment]
And of course now I need an issue tag to add to the test that I added to cover the framework feature that I added to cover the postprocessor that I added to test the kernels that I added to test the output I added to visualize the library feature I added to swallow the cat to catch the bird to catch the spider to catch the fly...

**Comment by lindsayad** [comment]
That was surprisingly straightforward to follow up to the cat :laughing: :grimacing: :cry:

**Review by lindsayad — APPROVED** [review]
Thanks for all the hard work on this

**Pull request merged**